### PR TITLE
qgschunkedentity: Use estimated gpu memory instead of hard coded limit

### PR DESCRIPTION
**Pull request by ptitjano**
## Description

This adds a routine to estimate the GPU available memory in kb. It uses some vendor OpenGL extensions. It works for AMD and Nvidia but there is no available extension for an Intel card. 
This estimation is then used in `QgsChunkedEntity` instead of the hard code memory limit.
This should mitigate the infinite 3d scene loading issue.

cc @benoitdm-oslandia @wonder-sk @uclaros

### CODE REVIEW

File: src/3d/chunks/qgschunkedentity_p.cpp [hunk 1]
```diff
@@ -62,6 +62,16 @@ QgsChunkedEntity::QgsChunkedEntity( float tau, QgsChunkLoaderFactory *loaderFact
...
+
+  int memoryAvailableKB = Qgs3DUtils::estimateGpuMemoryAvailable();
+  if ( memoryAvailableKB == -1 )
+  {
+    mGpuMemoryLimit = 500.0;
+  }
+  else
+  {
+    mGpuMemoryLimit = 0.8 * static_cast<double>( memoryAvailableKB ) / 1024.0;
+  }
```
Comment (nyalldawson) [inline src/3d/chunks/qgschunkedentity_p.cpp:74]: How about we leave the default 500mb value only defined in the header: (and add a comment next to it that it's a conservative default value for when the actual GPU available memory isn't available)
```suggestion
  if ( memoryAvailableKB != -1 )
  {
    mGpuMemoryLimit = 0.8 * static_cast<double>( memoryAvailableKB ) / 1024.0;
  }
```

**Comment by nyalldawson** [comment]
Nice work!

I think we could consider adding a user-visible setting for this too (eg in the Settings - Options - 3D tab). It could be a spin box which defaults to a clear value of "Automatic", but allow the user to manually set a desired value. That'd be useful for intel/apple/... users where the automatic value will be just the conservative 500mb.

**Review by wonder-sk — COMMENTED** [review]
Nice tweak, but the problem is that the mGpuMemoryLimit is for _each_ chunked entity, and there can be many in a single scene. So Giving 80% of GPU memory to each chunked entity is IMHO equivalent to no limit at all...

**Comment by ptitjano** [comment]
> Nice tweak, but the problem is that the mGpuMemoryLimit is for _each_ chunked entity, and there can be many in a single scene. So Giving 80% of GPU memory to each chunked entity is IMHO equivalent to no limit at all...

Does this mean the MemoryLimit should be moved to the map scene? Do you have some ideas?

**Comment by wonder-sk** [comment]
> Does this mean the MemoryLimit should be moved to the map scene? Do you have some ideas?

I don't know if moving to map scene would help: people can have more 3D scenes open at a time.
Moving the memory limit to QGIS wide option also may not be the solution: people can have multiple instances of QGIS open.

The more I think about it, I am not sure whether trying to base our limits automatically on the size of GPU memory is the correct approach. For example, I wouldn't like if QGIS used 80% of my RAM automatically, so not sure if it would make sense to use this approach for GPU either...

And there are various other GPU hungry processes like browsers...

I do not have a good solution.

**Pull request closed without merging**